ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Monad.hs

```diff
@@ -0,0 +1,126 @@
+{-# LANGUAGE BangPatterns      #-}
+{-# LANGUAGE DeriveFoldable    #-}
+{-# LANGUAGE DeriveFunctor     #-}
+{-# LANGUAGE DeriveTraversable #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE Rank2Types        #-}
+{-# LANGUAGE TypeFamilies      #-}
+
+-- | A monad for the interface to the ledger layer
+module Ouroboros.Consensus.Ledger.Monad (
+    -- * Ledger events
+    AuxLedgerEvent
+  , VoidLedgerEvent
+    -- * Ledger monad
+  , LedgerResult (..)
+  , LedgerT (..)
+  , coerceLedgerT
+  , evalLedgerT
+  , hoistLedgerT
+  , ledgerT
+  , liftLedgerT
+  , runLedgerT
+  , withLedgerT
+  ) where
+
+import           Control.Monad.Except
+import           Control.Monad.Morph (MFunctor (..))
+import           Control.Monad.Reader (ReaderT (ReaderT), runReaderT,
+                     withReaderT)
+import           Control.Monad.State.Strict (StateT (StateT), runStateT)
+import           Data.Functor ((<&>))
+import           Data.Kind (Type)
+
+{-------------------------------------------------------------------------------
+  Ledger monad
+-------------------------------------------------------------------------------}
+
+-- | Event emitted by the ledger
+--
+-- TODO we call this 'AuxLedgerEvent' to differentiate from 'LedgerEvent' in
+-- 'InspectLedger'. When that module is rewritten to make use of ledger
+-- derived events, we may rename this type.
+type family AuxLedgerEvent l :: Type
+
+-- | A 'Data.Void.Void' isomorph for ledgers that have no events.
+--
+data VoidLedgerEvent
+
+-- | The monad transformer used when interacting with the ledger
+newtype LedgerT l m a = LedgerT {
```

Comment:
Edsko suggests comparing this to the CPS-writer.

In particular, check `Control.Monad.Trans.Writer.CPS` -- is `mapWriterT` nice enough to use?

Reply:
Edsko and I decided to drop the monad from the interface.